### PR TITLE
Ensure unique x-coordinates with sampling (Fixes #20)

### LINKED ISSUE #20: Split method generating duplicate shares sometimes
**Description of the issue**
I need to use this module for sharing a secret key. To ensure the good working of it, I created a test function that generate a random key, split it in shares (3 parts and 2 thresholds), and then combining it back.

For some reason, it appears that sometimes I got the error [`Duplicate sample`](https://github.com/konidev20/pyshamir/blob/main/pyshamir/shamir.py#L39) when combining the keys. In fact, 2 of my keys are exactly the same.

After a bit of investigating, I found out that it is caused by [`x_coordinates`](https://github.com/konidev20/pyshamir/blob/main/pyshamir/shamir.py#L77) when generating the shares. If a number is duplicated in the `parts` firsts values (where `parts = 3` in my case), then the shares at index of those duplicates will be the same.

**Expected behaviour**
Shares are all different
good_behavior

**Random bad behaviour**
Some shares are exactly the same
bad_behavior

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to [`shamir.py` line 77](https://github.com/konidev20/pyshamir/blob/main/pyshamir/shamir.py#L77)
2. Replace `x_coordinates = [secrets.randbelow(255) for _ in range(1, 256)]` with some random numbers, except that there's a duplicate in the 3 firsts one. For example (notice the two `1` at the beginning of the list) : 
```py
x_coordinates = [1,1,57,119,170,112,72,53,195,195,131,192,11,101,31,200,190,241,209,89,228,99,82,125,78,208,199,62,225,97,136,19,152,115,66,53,245,214,98,78,165,113,182,159,8,64,223,201,118,117,246,207,22,136,48,50,218,73,107,43,219,83,188,234,104,1,108,206,221,38,183,77,96,138,35,1,213,184,246,241,242,64,239,246,37,206,22,224,159,133,221,64,207,211,24,128,114,227,236,96,42,254,165,64,103,233,143,45,225,97,187,140,98,174,183,209,132,225,233,94,64,180,200,109,190,40,143,153,210,207,220,130,139,75,114,87,45,45,93,111,102,42,247,50,82,208,233,52,67,242,193,218,158,239,22,220,47,136,114,99,244,150,39,198,21,108,64,228,180,214,215,24,248,81,193,215,6,106,84,128,26,196,232,9,247,119,203,64,247,158,16,45,233,108,85,67,94,39,157,64,240,184,129,151,158,117,85,105,254,116,203,252,160,90,56,226,118,105,157,106,239,162,39,50,117,187,187,102,177,101,114,96,250,70,103,26,113,168,2,140,5,151,105,136,129,235,167,75,154,199,204,243,149,156,251]
```
3. Run this simple piece of code (you can find it [here](https://www.mediafire.com/file/xtbgfza9hxo67z8/sss_error.zip/file)) :
```py
from base64 import b64decode
import pyshamir

secret = b64decode("a+m4G0kEkKDQK4MFGz7L0vLz5oViQkDSLThiC4zDRZU=")
shares = pyshamir.split(secret, 3, 2)
for share in shares:
    print(share.hex())
```
4. As a result, shares 1 and 2 will be exactly the same (as a result of the duplicates `1` in the list), and share 3 will be fine

**How to resolve**
I think that to solve this problem, the solution is as simple as ensure that the `parts` first numbers of this `x_coordinates` are unique. But as I don't know the deep working of this, I don't wanted to make a pull request.

**Desktop**
 - OS: Ubuntu 22.04
 - Python: 3.10.6
 - Pyshamir: 1.0.0

### PR DESCRIPTION
**Pull request by sidsbrmnn**
This pull request fixes an issue where duplicate x-coordinates could be generated. The fix involves using random sampling to ensure that all x-coordinates in the list are unique.

## Changes Made

- [pyshamir.py](pyshamir/pyshamir.py) - use `random.sample()` to generate a list of unique numbers than manually loop to generate a random secret.
- [test_pyshamir.py](tests/test_pyshamir.py) -  manually test that the parts are different

### CODE REVIEW

**Comment by sidsbrmnn** [comment]
Oops! Looks like we can't use `random.sample()`. Hmm, maybe just brute force the uniqueness.

**Comment by konidev20** [comment]
Can we make x_coordinates a set?

Also, thinking if generating random numbers per array index is a good idea in terms of security? 

Need to check the best way to generate a random array correctly.

**Comment by Loadingname91** [comment]
> Can we make x_coordinates a set?
> 
> Also, thinking if generating random numbers per array index is a good idea in terms of security?
> 
> Need to check the best way to generate a random array correctly.

How about using this

`import secrets

# Generate random list of x coordinates
x_coordinates = list(range(256))
secrets.SystemRandom().shuffle(x_coordinates)
x_coordinates = x_coordinates[:-1] # Remove last element to get list of 255 x coordinates`

File: pyshamir/shamir.py [hunk 1]
```diff
@@ -74,7 +74,11 @@ def split(secret: bytes, parts: int, threshold: int) -> list[bytearray]:
         raise ValueError("Secret must be at least 1 byte long")
 
     # Generate random list of x coordinates
-    x_coordinates = [secrets.randbelow(255) for _ in range(1, 256)]
+    x_coordinates = []
+    while len(x_coordinates) < 255:
+        x = secret.randbelow(255)
```
Comment (konidev20) [inline pyshamir/shamir.py:79]: this should be `secrets.randbelow(255)`. The unittests are failing but showing as successful. Need to check that in a separate issue.

**Review by konidev20 — APPROVED** [review]
LGTM. @Loadingname91 , @AbhishekSrikanth  please check and let me know.

File: pyshamir/_utils.py [hunk 2]
```diff
@@ -92,3 +92,9 @@ def interpolate_polynomial(x_samples, y_samples, x)->int:
         group = mul(y_samples[i], basis)
         result = add(result, group)
     return result
+
+def generate_x_coordinates(n):
+    x_coordinates = list(range(n+1))
+    secrets.SystemRandom().shuffle(x_coordinates)
+    x_coordinates = x_coordinates[:-1]
```
Comment (AbhishekSrikanth) [inline pyshamir/_utils.py:99]: Is it necessary to reverse this list?

**Pull request merged**